### PR TITLE
GPII-2507: Fix timing bug in gpii-testem "route" generation...

### PR DESCRIPTION
**Pull request by the-t-in-rtf**
See [GPII-2507](https://issues.gpii.net/browse/GPII-2507) for details.

### CODE REVIEW

**Comment by the-t-in-rtf** [comment]
@cindyli and @amb26, this fixes the timing bug that resulted in partial coverage reports when running the browser tests in infusion.  Ideally we would review this and cut a release before merging the related infusion pull.

**Comment by the-t-in-rtf** [comment]
@amb26, this one should be fairly self-contained, and it would be nice to cut a full "root safe" release.

**Comment by the-t-in-rtf** [comment]
In recent conversations on the [infusion pull that uses this code](https://github.com/fluid-project/infusion/pull/822), it's clear that this problem is not 100% resolved.  It's possible for instrumentation to silently fail without interrupting the overall run.  I will fix this shortly and we can review as part of this pull.

**Comment by the-t-in-rtf** [comment]
I fixed the immediate problem, but definitely need to migrate away from calling Istanbul using child_process.  See [GPII-2518](https://issues.gpii.net/browse/GPII-2518) for details.

**Comment by the-t-in-rtf** [comment]
This pull now includes the longer-term fix for consistently launching Istanbul from a dependent package.  It's ready for review.

File: yarn.lock [hunk 1]
```diff
@@ -23,14 +23,14 @@ acorn-jsx@^3.0.0:
...
 
-acorn@4.0.4:
-  version "4.0.4"
-  resolved "https://registry.yarnpkg.com/acorn/-/acorn-4.0.4.tgz#17a8d6a7a6c4ef538b814ec9abac2779293bf30a"
-
 acorn@^3.0.4, acorn@^3.1.0:
   version "3.3.0"
   resolved "https://registry.yarnpkg.com/acorn/-/acorn-3.3.0.tgz#45e37fb39e8da3f25baee3ff5369e2bb5f22017a"
 
+acorn@^5.0.1:
```
Comment (amb26) [inline yarn.lock]: This file will be axed, I imagine

File: src/js/testem-component.js [hunk 2]
```diff
@@ -340,36 +333,19 @@ gpii.testem.generateUniqueDirName = function (basePath, prefix, suffix) {
...
-                    promise.resolve();
-                }
-            });
+            var commandSegments = ["report", "--root", fluid.module.resolvePath(that.options.coverageDir), "--dir", that.options.reportsDir, "text-summary",  "html", "json-summary"];
+            cli.runToCompletion(commandSegments);
+            fluid.log("Created coverage report in '", that.options.reportsDir, "'...");
         }
         catch (error) {
-            promise.reject(error);
+            fluid.fail(error);
```
Comment (amb26) [inline src/js/testem-component.js]: What was the reason for converting this reject into a fail?

File: yarn.lock (hunk 1 above)
Comment (the-t-in-rtf) [reply yarn.lock]: This has already been addressed in the "headless" branch I'm working on.  I'm still wrestling a bit with Firefox, but I'll clean that up and submit just for Chrome once this PR is merged.

File: src/js/testem-component.js (hunk 2 above)
Comment (the-t-in-rtf) [reply src/js/testem-component.js]: My previous approach called istanbul as an external command asynchronously, hence the promise.  The current approach loads istanbul's `cli` module and calls that synchronously.

**Comment by amb26** [comment]
I get random test failures, it seems related to the erratic behaviour of IE11 windows which sometimes don't dismiss, or do dismiss and then trigger an extra failure. Here is a transcript:
```
21:39:20.702:  jq: Test concluded - Module "Testing coverage detection and reporting..." Test name "Running a suite of tests that results in complete coverage..
.": 6/6 passed - PASS
21:39:27.314:  TESTEM ERROR:{
    "killed": false,
    "code": 1,
    "signal": null,
    "cmd": "node ../node_modules/testem/testem.js ci --file testem-fixtures/testem-incomplete-coverage.js"
}
21:39:27.317:  jq: FAIL: Module "Testing coverage detection and reporting..." Test name "Running a suite of tests that results in incomplete coverage..." - Mess
age: There should be no errors running testem...
21:39:27.319:  jq: Source:     at pok (E:\Source\gits\gpii-testem\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:112:15)
    at Object.jsUnitCompat.fail (E:\Source\gits\gpii-testem\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:129:13)
    at E:\Source\gits\gpii-testem\tests\testem-component-tests.js:30:24
    at ChildProcess.exithandler (child_process.js:209:5)
    at emitTwo (events.js:106:13)
    at ChildProcess.emit (events.js:191:7)
    at maybeClose (internal/child_process.js:852:16)
    at Socket.<anonymous> (internal/child_process.js:323:11)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at Pipe._handle.close [as _onclose] (net.js:492:12)
21:39:27.323:  FATAL ERROR: Uncaught exception: C:\Users\Bosmon\AppData\Local\Temp\reports-33fxadm6-31\coverage-summary.json: Unexpected end of JSON input
SyntaxError: C:\Users\Bosmon\AppData\Local\Temp\reports-33fxadm6-31\coverage-summary.json: Unexpected end of JSON input
    at Object.parse (native)
    at Object.Module._extensions..json (module.js:558:27)
    at Module.load (module.js:458:32)
    at tryModuleLoad (module.js:417:12)
    at Function.Module._load (module.js:409:3)
    at Module.require (module.js:468:17)
    at require (internal/module.js:20:19)
    at E:\Source\gits\gpii-testem\tests\testem-component-tests.js:49:43
    at ChildProcess.exithandler (child_process.js:209:5)
    at emitTwo (events.js:106:13)
    at ChildProcess.emit (events.js:191:7)
    at maybeClose (internal/child_process.js:852:16)
    at Socket.<anonymous> (internal/child_process.js:323:11)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at Pipe._handle.close [as _onclose] (net.js:492:12)
21:39:27.327:  Current activity:
21:39:27.328:  jq: Test concluded - Module "Testing coverage detection and reporting..." Test name "Running a suite of tests that results in incomplete coverage
...": 5/6 passed - FAIL
21:39:32.952:  Removing dir 'C:\Users\Bosmon\AppData\Local\Temp\reports-7i4fedzh-31'...
21:39:32.996:  Removing dir 'C:\Users\Bosmon\AppData\Local\Temp\coverage-7i4fedzh-31'...
21:39:32.997:  Removed reports and coverage from this test run...
21:39:33.011:  jq: Test concluded - Module "Testing coverage detection and reporting..." Test name "Running a suite of tests without test coverage...": 4/4 pass
ed - PASS
21:39:44.904:  TESTEM ERROR:{
    "killed": false,
    "code": 1,
    "signal": null,
    "cmd": "node ../node_modules/testem/testem.js ci --file testem-fixtures/testem-instrumentation-timing.js"
}
21:39:44.912:  jq: FAIL: Module "Testing coverage detection and reporting..." Test name "Confirm that long-running instrumentation does not interfere with cover
age collection..." - Message: There should be no errors running testem...
21:39:44.913:  jq: Source:     at pok (E:\Source\gits\gpii-testem\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:112:15)
    at Object.jsUnitCompat.fail (E:\Source\gits\gpii-testem\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:129:13)
    at E:\Source\gits\gpii-testem\tests\testem-component-tests.js:30:24
    at ChildProcess.exithandler (child_process.js:209:5)
    at emitTwo (events.js:106:13)
    at ChildProcess.emit (events.js:191:7)
    at maybeClose (internal/child_process.js:852:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:215:5)
21:39:44.918:  FATAL ERROR: Uncaught exception: C:\Users\Bosmon\AppData\Local\Temp\reports-81enn9jq-32\coverage-summary.json: Unexpected end of JSON input
SyntaxError: C:\Users\Bosmon\AppData\Local\Temp\reports-81enn9jq-32\coverage-summary.json: Unexpected end of JSON input
    at Object.parse (native)
    at Object.Module._extensions..json (module.js:558:27)
    at Module.load (module.js:458:32)
    at tryModuleLoad (module.js:417:12)
    at Function.Module._load (module.js:409:3)
    at Module.require (module.js:468:17)
    at require (internal/module.js:20:19)
    at E:\Source\gits\gpii-testem\tests\testem-component-tests.js:49:43
    at ChildProcess.exithandler (child_process.js:209:5)
    at emitTwo (events.js:106:13)
    at ChildProcess.emit (events.js:191:7)
    at maybeClose (internal/child_process.js:852:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:215:5)
21:39:44.926:  Current activity:
21:39:44.927:  jq: Test concluded - Module "Testing coverage detection and reporting..." Test name "Confirm that long-running instrumentation does not interfere
 with coverage collection...": 5/6 passed - FAIL
21:39:47.676:  FATAL ERROR: Uncaught exception: Path must be a string. Received null
TypeError: Path must be a string. Received null
    at assertPath (path.js:7:11)
    at Object.resolve (path.js:184:7)
    at E:\Source\gits\gpii-testem\tests\testem-component-tests.js:38:42
    at ChildProcess.exithandler (child_process.js:209:5)
    at emitTwo (events.js:106:13)
    at ChildProcess.emit (events.js:191:7)
    at maybeClose (internal/child_process.js:852:16)
    at Socket.<anonymous> (internal/child_process.js:323:11)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at Pipe._handle.close [as _onclose] (net.js:492:12)
21:39:47.680:  Current activity:
21:39:47.687:  jq: Test concluded - Module "Testing coverage detection and reporting..." Test name "Running a suite of tests with gross configuration errors..."
: 2/2 passed - PASS
21:39:47.692:  jq: ***************
21:39:47.693:  jq: All tests concluded: 30/32 total passed in 56128ms - FAIL
21:39:47.694:  jq: ***************
```

The final failure, "Path must be a string. Received null" appears consistent in each run though.

Windows 7 x64, with IE11 installed

**Comment by the-t-in-rtf** [comment]
> The final failure, "Path must be a string. Received null" appears consistent in each run though.

That's a "failure" test, the exception is expected and correct.  I cannot reproduce the rest in Windows 10, but will keep investigating.

**Comment by amb26** [comment]
Yes, I can't reproduce in the Windows 10 vagrant box either. The failures all seem to be due to the IE11 windows which don't close. We could decide to put IE11 out of scope, but this then means that this project can't be used to test Infusion for which I believe IE11 is still in the supported list.

**Comment by the-t-in-rtf** [comment]
@amb26, the tests do pass on IE11 on the Windows 10 VM.  We need to confirm whether the tests fail on the Windows 7 VM, and then discuss further.

**Comment by the-t-in-rtf** [comment]
While waiting for the VM to build I have been looking through Testem issues and found this, which is supposed to have been fixed, but which sounds familiar:

https://github.com/testem/testem/issues/583

**Comment by the-t-in-rtf** [comment]
Thankfully, even though it's far out of date, we do still have a Windows 7 VM (inclusivedesign/windows7-eval-x64-Apps).  Before you can run `vagrant ci test` against that VM, you have to manually mount v:, as in `net use v: \\vboxsrv\vagrant`.   For your convenience, I have committed a copy of the `.vagrant.yml` file with the relevant box and commands commented out.

When running `npm test` in that VM, the IE11 windows all close.  However, I noticed that I had to hit ctrl+c at the end of the overall test run after the last test before the coverage report appears.  I looked into this further, and confirmed that it happens whether or not I use `nyc`.

Moving on to running individual tests, I discovered that the failure tests fail beyond the way in which they are intended to, and break part of the test harness itself.  It appears that I see the same error on OS X, at least when running the test directly.  I will investigate further, I suspect that we'll need to improve the failure mode checks for expected failure and then refine the failure test.  I will comment further when that's done.

In investigating, I did bump the versions on `nyc`, `testem` and other dependencies.  I am not able to reproduce the browser closure issues at all.  Hold off for a bit, I'll comment when the failure modes are no longer "not even wrong".

**Comment by the-t-in-rtf** [comment]
Just to close the loop, per comments @amb26 made in a chat, I confirmed that Testem itself fails to close IE11 windows if IE is already open when you start the test run.  I filed a bug here:

https://github.com/testem/testem/issues/1184

I will add a note to the docs shortly.

**Comment by the-t-in-rtf** [comment]
OK, @amb26, this should be ready to review again.

**Comment by the-t-in-rtf** [comment]
Just to confirm, @amb26 and I are working through some unique slowness when running with "headless" Chrome which appears in his environment.

**Comment by the-t-in-rtf** [comment]
I have updated so that "headless" is disabled by default, but can be enabled using the `HEADLESS` environment variable.

**Comment by the-t-in-rtf** [comment]
In working on the similar concerns around gpii-webdriver, I realised that when attempting to run in "headless" mode we should fail over to "headed" options if there are no "headless" options for a given browser.  I updated this work to reflect that.  Just to confirm, @amb26, as far as I know this should be ready for further review.

**Comment by the-t-in-rtf** [comment]
@amb26, I am currently thinking through a new spike in this project to add better instrumentation and test cases that more closely mirror complex scenarios like we see with universal.  I was hoping you could close this one out before I begin that work, it's been ready for review for a while.

File: tests/harness.js [hunk 3]
```diff
@@ -52,6 +52,38 @@ fluid.defaults("gpii.tests.testem.harness", {
             priority: "before:cleanup",
             funcName: "gpii.tests.testem.harness.outputOptions",
             args:     ["{that}"]
+        },
+        constructFixtures: {
```
Comment (amb26) [inline tests/harness.js]: Let's namespace all of this set - e.g. with names such as "constructFixtures.log", etc.

Comment (the-t-in-rtf) [reply tests/harness.js]: Done.

**Comment by the-t-in-rtf** [comment]
Thanks, @amb26.  I have cut the 1.0.2 release and updated the forward-facing version in master.

**Pull request merged**